File: solutions/c/reverse-string/1/reverse_string.c
#include "reverse_string.h"
#include <stdlib.h>
#include <string.h>

char *reverse(const char *value) {
    if (value == NULL) return NULL;
    
    size_t len = strlen(value);
    char *result = malloc(len + 1);
    if (result == NULL) return NULL;
    
    for (size_t i = 0; i < len; ++i) {
        result[i] = value[len - 1 - i];
    }
    result[len] = '\0';
    
    return result;
}

File: solutions/fortran/reverse-string/1/reverse_string.f90
module reverse_string
  implicit none
contains

  function reverse(input) result(reversed)
    character(*), intent(in) :: input
    character(len=len(input)) :: reversed
    integer :: i, n

    n = len(input)
    do i = 1, n
      reversed(i:i) = input(n-i+1:n-i+1)
    end do
  end function

end module